src/beamlime/core/application.py

```diff
@@ -40,19 +40,48 @@ def __del__(self) -> None:
         ...
 
 
-class BeamlimeApplicationInterface(ABC, BeamLimeApplicationProtocol):
+class _LogMixin:
+    def _log(self, level: int = DEBUG, msg="", *args):
+        self.logger._log(level, msg={"app_name": self.app_name, "msg": msg}, args=args)
+
+    def debug(self, msg: str) -> None:
+        self._log(level=DEBUG, msg=msg)
+
+    def info(self, msg: str) -> None:
+        self._log(level=INFO, msg=msg)
+
+    def warn(self, msg: str) -> None:
+        self._log(level=WARN, msg=msg)
+
+    def exception(self, msg: str) -> None:
+        self._log(level=ERROR, msg=msg)
+
+    def error(self, msg: str) -> None:
+        self._log(level=ERROR, msg=msg)
+
+
+class BeamlimeApplicationInterface(_LogMixin, ABC, BeamLimeApplicationProtocol):
```

Comment:
Didn't notice earlier, but why does this inherit from `BeamlimeApplicationProtocol`?

Reply:
I wanted to keep minimum abstrace method in the `Protocol` while we have common methods in the `Interface`.
In principle it doesn't have to inherit but it was easier to check if it to has all methods and properties that the Protocol require...? Should it not inherit from the Protocol...?

Reply:
AFAIK, the purpose of protocols is to define an interface without requiring inheritance. If you want to enforce compliance with the interface, you should just use `abc` and no protocols.